File: host-windows/src/AppControl.Host/Security/SecureChannel.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;
using NSec.Cryptography;

namespace AppControl.Host.Security;

public enum ChannelPhase : byte { Handshake = 0x01, Session = 0x02 }
public enum ChannelDirection : byte { HostToViewer = 0x01, ViewerToHost = 0x02 }

/// <summary>
/// Ende-zu-Ende-Verschluesselung des Signalings.
///
/// Diese Klasse ist das C#-Gegenstueck zu Security/SecureChannel.swift im Viewer
/// und zu tools/crypto-vectors/generate.mjs. Alle drei muessen bitgleiche
/// Ergebnisse liefern; CryptoVectorTests.cs prueft das gegen vectors.json.
///
/// Siehe docs/04-protocol.md §4.2 und docs/05-security.md §5.3.
/// </summary>
public sealed class SecureChannel
{
    public const int HeaderSize = 20;
    public const int KeySize = 32;
    private static readonly byte[] Magic = "AC1\0"u8.ToArray();

    private readonly Key _key;
    private readonly ChannelDirection _sendDirection;
    private readonly ChannelPhase _phase;
    private readonly byte[] _sendSalt = new byte[4];

    private ulong _sendCounter;
    private ulong _highestReceivedCounter;

    public SecureChannel(ReadOnlySpan<byte> keyMaterial, ChannelPhase phase, ChannelDirection sendDirection)
    {
        if (keyMaterial.Length != KeySize)
            throw new ArgumentException($"Schluessel muss {KeySize} Byte lang sein", nameof(keyMaterial));

        _key = Key.Import(AeadAlgorithm.ChaCha20Poly1305, keyMaterial, KeyBlobFormat.RawSymmetricKey);
        _phase = phase;
        _sendDirection = sendDirection;
        RandomNumberGenerator.Fill(_sendSalt);
    }

    private ChannelDirection ReceiveDirection => _sendDirection == ChannelDirection.HostToViewer
        ? ChannelDirection.ViewerToHost
        : ChannelDirection.HostToViewer;

    /// <summary>Verschluesselt einen Klartext zu einem sendefertigen Umschlag.</summary>
    public byte[] Seal(ReadOnlySpan<byte> plaintext)
    {
        var counter = Interlocked.Increment(ref Unsafe_Counter());
        var header = BuildHeader(_phase, _sendDirection, counter, _sendSalt);

        var envelope = new byte[HeaderSize + plaintext.Length + 16];
        header.CopyTo(envelope);

        // Nonce sind die Bytes 8..20 des Headers: Zaehler(8, big-endian) + Salt(4).
        // Der Zaehler steigt streng monoton, der Salt ist pro Kanal zufaellig -
        // damit ist eine Nonce-Wiederverwendung ueber Sitzungen hinweg
        // ausgeschlossen, was bei ChaCha20-Poly1305 katastrophal waere.
        var nonce = new ReadOnlySpan<byte>(header, 8, 12);

        AeadAlgorithm.ChaCha20Poly1305.Encrypt(
            _key, nonce, associatedData: header, plaintext,
            envelope.AsSpan(HeaderSize));

        return envelope;
    }

    /// <summary>
    /// Entschluesselt einen Umschlag. Gibt null zurueck bei fehlerhaftem Format,
    /// falscher Richtung, Replay oder gescheiterter Authentifizierung.
    ///
    /// Bewusst KEINE Exception: Ein manipulierter Umschlag ist ein normaler
    /// Betriebszustand bei einer Verbindung ueber das offene Internet. Der
    /// Aufrufer zaehlt Fehlversuche und trennt bei Haeufung
    /// (docs/04-protocol.md §4.8).
    /// </summary>
    public byte[]? Open(ReadOnlySpan<byte> envelope)
    {
        if (envelope.Length < HeaderSize + 16) return null;

        var header = envelope[..HeaderSize];

        if (!header[..4].SequenceEqual(Magic)) return null;
        if ((ChannelPhase)header[4] != _phase) return null;

        // Richtungspruefung: Ohne sie koennte ein Angreifer eine an uns gesendete
        // Nachricht zurueckspiegeln, und wir wuerden sie als Antwort des Peers
        // akzeptieren (Reflection-Angriff).
        if ((ChannelDirection)header[5] != ReceiveDirection) return null;

        var counter = BinaryPrimitives.ReadUInt64BigEndian(header[8..16]);

        // Replay-Schutz: Der Zaehler muss echt groesser sein als alles bisher
        // Akzeptierte. Ein wiederholt eingespielter Umschlag wird verworfen.
        if (counter <= _highestReceivedCounter) return null;

        var nonce = header[8..20];
        var ciphertext = envelope[HeaderSize..];
        var plaintext = new byte[ciphertext.Length - 16];

        if (!AeadAlgorithm.ChaCha20Poly1305.Decrypt(
                _key, nonce, associatedData: header, ciphertext, plaintext))
        {
            return null;
        }

        // Erst NACH erfolgreicher Authentifizierung fortschreiben. Sonst koennte
        // ein Angreifer mit gefaelschten hohen Zaehlern legitime Nachrichten
        // dauerhaft aussperren.
        _highestReceivedCounter = counter;
        return plaintext;
    }

    internal static byte[] BuildHeader(ChannelPhase phase, ChannelDirection direction,
                                       ulong counter, ReadOnlySpan<byte> salt4)
    {
        var header = new byte[HeaderSize];
        Magic.CopyTo(header, 0);
        header[4] = (byte)phase;
        header[5] = (byte)direction;
        // header[6..8] reserviert, bleibt 0
        BinaryPrimitives.WriteUInt64BigEndian(header.AsSpan(8, 8), counter);
        salt4[..4].CopyTo(header.AsSpan(16, 4));
        return header;
    }

    private ref long Unsafe_Counter()
    {
        // Interlocked braucht ein long-Feld; der Zaehler ist logisch ulong.
        // Bei 2^63 Nachrichten waere das ein Problem - das sind bei 1000/s
        // etwa 290 Millionen Jahre.
        return ref System.Runtime.CompilerServices.Unsafe.As<ulong, long>(ref _sendCounter);
    }

    /// <summary>Testhilfe: erlaubt deterministisches Versiegeln fuer Testvektoren.</summary>
    internal byte[] SealWithFixedNonce(ReadOnlySpan<byte> plaintext, ulong counter, ReadOnlySpan<byte> salt4)
    {
        var header = BuildHeader(_phase, _sendDirection, counter, salt4);
        var envelope = new byte[HeaderSize + plaintext.Length + 16];
        header.CopyTo(envelope, 0);
        AeadAlgorithm.ChaCha20Poly1305.Encrypt(
            _key, new ReadOnlySpan<byte>(header, 8, 12), header, plaintext, envelope.AsSpan(HeaderSize));
        return envelope;
    }
}

File: viewer-macos/Sources/AppControlViewer/Security/SecureChannel.swift
import Crypto
import Foundation

enum ChannelPhase: UInt8 { case handshake = 0x01, session = 0x02 }
enum ChannelDirection: UInt8 { case hostToViewer = 0x01, viewerToHost = 0x02 }

/// Ende-zu-Ende-Verschlüsselung des Signalings.
///
/// Swift-Gegenstück zu `Security/SecureChannel.cs` und
/// `tools/crypto-vectors/generate.mjs`. Alle drei müssen bitgleiche Ergebnisse
/// liefern; `CryptoVectorTests.swift` prüft das gegen `vectors.json`.
///
/// Siehe docs/04-protocol.md §4.2.
final class SecureChannel {
    static let headerSize = 20
    static let keySize = 32
    private static let magic: [UInt8] = Array("AC1\0".utf8)

    private let key: SymmetricKey
    private let phase: ChannelPhase
    private let sendDirection: ChannelDirection
    private let sendSalt: Data

    private var sendCounter: UInt64 = 0
    private var highestReceivedCounter: UInt64 = 0
    private let lock = NSLock()

    init(key keyMaterial: Data, phase: ChannelPhase, sendDirection: ChannelDirection) {
        precondition(keyMaterial.count == Self.keySize, "Schlüssel muss 32 Byte lang sein")
        self.key = SymmetricKey(data: keyMaterial)
        self.phase = phase
        self.sendDirection = sendDirection

        var salt = Data(count: 4)
        salt.withUnsafeMutableBytes { buffer in
            _ = SecRandomCopyBytes(kSecRandomDefault, 4, buffer.baseAddress!)
        }
        self.sendSalt = salt
    }

    private var receiveDirection: ChannelDirection {
        sendDirection == .hostToViewer ? .viewerToHost : .hostToViewer
    }

    /// Verschlüsselt einen Klartext zu einem sendefertigen Umschlag.
    func seal(_ plaintext: Data) throws -> Data {
        lock.lock()
        sendCounter += 1
        let counter = sendCounter
        lock.unlock()

        return try Self.seal(key: key, phase: phase, direction: sendDirection,
                             counter: counter, salt4: sendSalt, plaintext: plaintext)
    }

    /// Interne Variante mit festem Nonce — für Testvektoren.
    static func seal(key: SymmetricKey, phase: ChannelPhase, direction: ChannelDirection,
                     counter: UInt64, salt4: Data, plaintext: Data) throws -> Data {
        let header = buildHeader(phase: phase, direction: direction, counter: counter, salt4: salt4)

        // Nonce sind die Bytes 8..20 des Headers: Zähler(8, big-endian) + Salt(4).
        // Der Zähler steigt streng monoton, der Salt ist pro Kanal zufällig —
        // damit ist eine Nonce-Wiederverwendung ausgeschlossen, die bei
        // ChaCha20-Poly1305 katastrophal wäre.
        let nonce = try ChaChaPoly.Nonce(data: header[8..<20])

        let box = try ChaChaPoly.seal(plaintext, using: key, nonce: nonce, authenticating: header)

        var envelope = Data(header)
        envelope.append(box.ciphertext)
        envelope.append(box.tag)
        return envelope
    }

    /// Entschlüsselt einen Umschlag.
    ///
    /// Gibt `nil` zurück bei fehlerhaftem Format, falscher Richtung, Replay oder
    /// gescheiterter Authentifizierung. Bewusst **kein** `throws`: Ein
    /// manipulierter Umschlag ist ein normaler Betriebszustand bei einer
    /// Verbindung über das offene Internet. Der Aufrufer zählt Fehlversuche und
    /// trennt bei Häufung (docs/04-protocol.md §4.8).
    func open(_ envelope: Data) -> Data? {
        guard envelope.count >= Self.headerSize + 16 else { return nil }

        let header = envelope.prefix(Self.headerSize)
        let bytes = [UInt8](header)

        guard Array(bytes[0..<4]) == Self.magic else { return nil }
        guard bytes[4] == phase.rawValue else { return nil }

        // Richtungsprüfung: Ohne sie könnte ein Angreifer eine an uns gesendete
        // Nachricht zurückspiegeln, und wir würden sie als Antwort des Peers
        // akzeptieren (Reflection-Angriff).
        guard bytes[5] == receiveDirection.rawValue else { return nil }

        let counter = bytes[8..<16].reduce(UInt64(0)) { ($0 << 8) | UInt64($1) }

        // Replay-Schutz: Der Zähler muss echt größer sein als alles bisher
        // Akzeptierte.
        lock.lock()
        let lastAccepted = highestReceivedCounter
        lock.unlock()
        guard counter > lastAccepted else { return nil }

        let ciphertext = envelope[(envelope.startIndex + Self.headerSize)..<(envelope.endIndex - 16)]
        let tag = envelope.suffix(16)

        do {
            let nonce = try ChaChaPoly.Nonce(data: header[(header.startIndex + 8)..<(header.startIndex + 20)])
            let box = try ChaChaPoly.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
            let plaintext = try ChaChaPoly.open(box, using: key, authenticating: Data(header))

            // Erst NACH erfolgreicher Authentifizierung fortschreiben. Sonst
            // könnte ein Angreifer mit gefälschten hohen Zählern legitime
            // Nachrichten dauerhaft aussperren.
            lock.lock()
            highestReceivedCounter = counter
            lock.unlock()

            return plaintext
        } catch {
            return nil
        }
    }

    static func buildHeader(phase: ChannelPhase, direction: ChannelDirection,
                            counter: UInt64, salt4: Data) -> Data {
        var header = Data(count: headerSize)
        header.replaceSubrange(0..<4, with: magic)
        header[4] = phase.rawValue
        header[5] = direction.rawValue
        // header[6..8] reserviert, bleibt 0
        withUnsafeBytes(of: counter.bigEndian) { header.replaceSubrange(8..<16, with: $0) }
        header.replaceSubrange(16..<20, with: salt4.prefix(4))
        return header
    }
}
